stl/inc/utility

```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !_Is_any_of_v<_Ty, bool, char,
+#ifdef _NATIVE_WCHAR_T_DEFINED
+        wchar_t,
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// FUNCTION TEMPLATE cmp_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    if constexpr (is_signed_v<_Ty1> == is_signed_v<_Ty2>) {
+        return _Left == _Right;
+    } else if constexpr (is_signed_v<_Ty2>) {
+        return _Left == static_cast<make_unsigned_t<_Ty2>>(_Right) && _Right >= 0;
+    } else {
+        return static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right && _Left >= 0;
+    }
+}
+
+// FUNCTION TEMPLATE cmp_not_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !cmp_equal(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE cmp_less
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_less(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    if constexpr (is_signed_v<_Ty1> == is_signed_v<_Ty2>) {
+        return _Left < _Right;
+    } else if constexpr (is_signed_v<_Ty2>) {
+        return _Right > 0 && _Left < static_cast<make_unsigned_t<_Ty2>>(_Right);
+    } else {
+        return _Left < 0 || static_cast<make_unsigned_t<_Ty1>>(_Left) < _Right;
+    }
+}
+
+// FUNCTION TEMPLATE cmp_greater
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_greater(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return cmp_less(_Right, _Left);
+}
+
+// FUNCTION TEMPLATE cmp_less_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_less_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !cmp_less(_Right, _Left);
+}
+
+// FUNCTION TEMPLATE cmp_greater_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !cmp_less(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE in_range
+template <class _Ty1, class _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    using _UTy1 = make_unsigned_t<_Ty1>;
+    // _U_max is numeric_limits<_Ty1>::max() converted to _UTy1
+    constexpr auto _U_max = static_cast<_UTy1>(static_cast<_UTy1>(-1) >> is_signed_v<_Ty1>);

```

Comment:
I wonder if this should contain `constexpr` programming to return guaranteed answers when we can statically see that the bounds are safe. When `_Ty1`'s range is a (non-strict) superset of `_Ty2`'s range (including when they're the same type), we can always return `true`. We can also skip bounds checks in some cases - e.g. when `is_signed_v<_Ty2>` but `_Ty1` is unsigned, the code below performs two checks, but at constexpr time we can detect if `_Ty1`'s upper bound is plenty - e.g. `_Ty2` could be `int16_t` and `_Ty1` could be `uint64_t` so we only need the `>= 0` check.

So, I think I'm asking for a significant refactoring. I haven't worked out what the code should look like, but I believe it would consider each bound (lower/upper) separately, and use constexpr tests to determine if they're always satisfied, or if we need a runtime comparison, leading to `return true;` in cases where the bounds are definitely ok.

This would also avoid compiler warnings when comparisons are always true.

Reply:
Also, if such logic would be really obnoxious, then it's fine to not do this - I'm just wondering whether it's possible to avoid compiler warnings and improve debug codegen at minimal cost.

Reply:
I figured out a really elegant way of expressing this. Will push a change.